### ISSUE TITLE
magit failed to build with borg--native-compile

### ISSUE THREAD

**Issue by Eason0210** [magit/magit#4665]
Hi @tarsius 

I update magit to latest commit.  But got these error, when I run` make clean all`.  
Note: I have this config in` ~/.emacs.d/etc/borg/config.el` and I have already install the `compat.el`
```emacs-lisp
(when (native-comp-available-p)
  (setq borg-compile-function #'native-compile))
```
```bash
--- [magit] ---

  Loading /Users/eason/.emacs.d/etc/borg/config.el...
 Creating /Users/eason/.emacs.d/lib/magit/lisp/magit-autoloads.el...
 Skipping /Users/eason/.emacs.d/lib/magit/lisp/git-commit-pkg.el...skipped
Compiling /Users/eason/.emacs.d/lib/magit/lisp/git-commit.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/git-commit.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/git-rebase.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/git-rebase.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-apply.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-apply.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-autorevert.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-autorevert.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-base.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-base.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-bisect.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-bisect.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-blame.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-blame.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-bookmark.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-bookmark.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-branch.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-branch.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-bundle.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-bundle.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-clone.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-clone.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-commit.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-commit.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-core.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-core.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-diff.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-diff.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-ediff.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-ediff.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-extras.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-extras.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-fetch.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-fetch.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-files.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-files.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-git.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-git.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-gitignore.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-gitignore.el" "No version number could be extracted")
 Skipping /Users/eason/.emacs.d/lib/magit/lisp/magit-libgit-pkg.el...skipped
 Skipping /Users/eason/.emacs.d/lib/magit/lisp/magit-libgit.el...skipped
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-log.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-log.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-margin.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-margin.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-merge.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-merge.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-mode.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-mode.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-notes.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-notes.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-obsolete.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-obsolete.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-patch.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-patch.el" "No version number could be extracted")
 Skipping /Users/eason/.emacs.d/lib/magit/lisp/magit-pkg.el...skipped
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-process.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-process.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-pull.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-pull.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-push.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-push.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-reflog.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-reflog.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-refs.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-refs.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-remote.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-remote.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-repos.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-repos.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-reset.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-reset.el" "No version number could be extracted")
 Skipping /Users/eason/.emacs.d/lib/magit/lisp/magit-section-pkg.el...skipped
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-section.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-section.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-sequence.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-sequence.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-sparse-checkout.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-sparse-checkout.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-stash.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-stash.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-status.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-status.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-submodule.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-submodule.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-subtree.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-subtree.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-tag.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-tag.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-transient.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-transient.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-wip.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-wip.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit-worktree.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit-worktree.el" "No version number could be extracted")
Compiling /Users/eason/.emacs.d/lib/magit/lisp/magit.el...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
borg--native-compile: (error "/Users/eason/.emacs.d/lib/magit/lisp/magit.el" "No version number could be extracted")
Done (Total of 0 files compiled, 47 failed, 5 skipped)
  Running ‘makeinfo --no-split magit-section.texi -o magit-section.info’...
  Running ‘makeinfo --no-split magit-section.texi -o magit-section.info’...done
  Running ‘makeinfo --no-split magit.texi -o magit.info’...
  Running ‘makeinfo --no-split magit.texi -o magit.info’...done
  Running ‘install-info magit-section.info --dir=dir’...
  Running ‘install-info magit-section.info --dir=dir’...done
  Running ‘install-info magit.info --dir=dir’...
  Running ‘install-info magit.info --dir=dir’...done
```
It works well before [magit 3cfc8458 ](https://github.com/magit/magit/commit/3cfc8458)

Platform: macOS 12.3.1
Emacs version : emacs 29.0.50 master[ commit 132fa71](https://github.com/emacs-mirror/emacs/commit/132fa716155ca1d4faa0ae936d14e7831e88dcf5)
borg version: [latest commit a968c5e](https://github.com/emacscollective/borg/commit/a968c5e)

**Comment by tarsius** [comment]
Is `compat` up-to-date as well? And since you mention native compilation, does it work without that?

**Comment by tarsius** [comment]
Try adding debug statements to `compat--generate-verbose` to find out more.

**Comment by Eason0210** [comment]
>  Is `compat` up-to-date as well? 
The` compat.el  i`s up-to-date.  I am using the [commit 401df6d](https://github.com/emacsmirror/compat/commit/401df6d)

>And since you mention native compilation, does it work without that?

I try again without native-compilation, It works well. 
So the issue is cause by borg.

**Comment by Eason0210** [comment]
> Try adding debug statements to `compat--generate-verbose` to find out more.
Update compat to use   `compat--generate-verbose`  and build it again with `M-x borg-build , compat`
```bash
(01:33:07) Loading /Users/eason/.emacs.d/etc/borg/config.el

(01:33:07) Building compat

Initializing drones...done (92 drones in 0.225s)
Loading /Users/eason/.emacs.d/etc/borg/config.el...
Creating /Users/eason/.emacs.d/lib/compat/compat-autoloads.el...
Compiling /Users/eason/.emacs.d/lib/compat/compat-24.el...

In compat--bool-vector-count-consecutive:
lib/compat/compat-24.el:219:2: Warning: docstring wider than 80 characters

In compat--bool-vector-count-population:
lib/compat/compat-24.el:233:2: Warning: docstring wider than 80 characters

In compat--completion-table-with-cache:
lib/compat/compat-24.el:350:2: Warning: docstring wider than 80 characters
Compiling /Users/eason/.emacs.d/lib/compat/compat-25.el...
Compiling /Users/eason/.emacs.d/lib/compat/compat-26.el...

In compat--file-attribute-modification-time:
lib/compat/compat-26.el:463:2: Warning: docstring wider than 80 characters

In compat--file-attribute-status-change-time:
lib/compat/compat-26.el:470:2: Warning: docstring wider than 80 characters
Compiling /Users/eason/.emacs.d/lib/compat/compat-27.el...
Compiling /Users/eason/.emacs.d/lib/compat/compat-28.el...

In compat--file-modes-number-to-symbolic:
lib/compat/compat-28.el:620:2: Warning: docstring wider than 80 characters
Compiling /Users/eason/.emacs.d/lib/compat/compat-font-lock.el...
Compiling /Users/eason/.emacs.d/lib/compat/compat-help.el...
Compiling /Users/eason/.emacs.d/lib/compat/compat-macs.el...
Skipping /Users/eason/.emacs.d/lib/compat/compat-tests.el...skipped
Compiling /Users/eason/.emacs.d/lib/compat/compat.el...
Done (Total of 9 files compiled, 1 skipped)
Running ‘makeinfo --no-split compat.texi -o compat.info’...
Running ‘makeinfo --no-split compat.texi -o compat.info’...done
Running ‘install-info compat.info --dir=dir’...
Running ‘install-info compat.info --dir=dir’...done

Process emacs ... --eval (borg-build "compat") finished
```
And then build magit again, got the same error:
``` bash

(01:37:54) Loading /Users/eason/.emacs.d/etc/borg/config.el

(01:37:54) Building magit

Initializing drones...done (92 drones in 0.157s)
Loading /Users/eason/.emacs.d/etc/borg/config.el...
Creating /Users/eason/.emacs.d/lib/magit/lisp/magit-autoloads.el...
Skipping /Users/eason/.emacs.d/lib/magit/lisp/git-commit-pkg.el...skipped
Compiling /Users/eason/.emacs.d/lib/magit/lisp/git-commit.el...
Loading /Users/eason/.emacs.d/lib/compat/compat-24.el (source)...
Loading /Users/eason/.emacs.d/lib/compat/compat-25.el (source)...
Eager macro-expansion failure: (error "No version number could be extracted")
Eager macro-expansion failure: (error "No version number could be extracted")
Error: /Users/eason/.emacs.d/lib/magit/lisp/git-commit.el: "No version number could be extracted"
Done (Total of 0 files compiled, 1 skipped)
Running ‘makeinfo --no-split magit-section.texi -o magit-section.info’...
Running ‘makeinfo --no-split magit-section.texi -o magit-section.info’...done
Running ‘makeinfo --no-split magit.texi -o magit.info’...
Running ‘makeinfo --no-split magit.texi -o magit.info’...done
Running ‘install-info magit-section.info --dir=dir’...
Running ‘install-info magit-section.info --dir=dir’...done
Running ‘install-info magit.info --dir=dir’...
Running ‘install-info magit.info --dir=dir’...done

Process emacs ... --eval (borg-build "magit") finished
```

**Comment by Eason0210** [comment]
With magit bitsect, I got the following info, may it would be help:

```
571b4346c1057cd504ddb6497c6b35c5d38eaa61 is the first bad commit
commit 571b4346c1057cd504ddb6497c6b35c5d38eaa61
Author: Jonas Bernoulli <jonas@bernoul.li>
Date:   Fri Apr 22 18:14:52 2022 +0200
    Depend on the compat package from GNU Elpa
 default.mk                | 11 +++++++++++
 lisp/git-commit-pkg.el    |  1 +
 lisp/git-commit.el        |  2 +-
 lisp/magit-base.el        |  3 +++
 lisp/magit-libgit-pkg.el  |  1 +
 lisp/magit-libgit.el      |  5 ++++-
 lisp/magit-pkg.el         |  1 +
 lisp/magit-section-pkg.el |  1 +
 lisp/magit-section.el     |  5 ++++-
 lisp/magit.el             |  2 +-
 10 files changed, 28 insertions(+), 4 deletions(-)
```

**Comment by tarsius** [comment]
Yes, the problem obviously occurs when we use `compat`, so until we use `compat` there is no problem.

Even when merely byte-compiling, `compat` has to be compiled before any package is compiled that depends on it. I believe that can be improved but I haven't looked into it in detail yet. Currently `borg` deals with it by making sure that `compat` is byte-compiled first. Maybe there is a similar issue with native-compilation. I will look into that too, though probably not today (`compat`'s author isn't around until Monday anyway).

**Comment by tarsius** [comment]
Please try to reproduce this without `borg` and report it at https://todo.sr.ht/~pkal/compat.

If this issue lived in `borg`'s issue tracker, I would leave it open even though I am guessing it has to be addressed in `compat`. Since I cannot transfer this issue across organization borders, I am closing it, since it isn't a Magit issue. We can still continue to discuss here but `compat`'s issue tracker is a more suitable place.

**Comment by Eason0210** [comment]
I agree that the issue is  from `compat`, but  I have no idea how to reproduce it without`borg`.

With latest commit of  `borg`, when I run `make all`, `borg` will be the first package to execute `borg--native-compile` , and then `compat` is the second one.

But when compile other packages( such as `magit`, `org-roam` , etc ...) which depends on `compat`,  will cause the issue.  I found that other packages need to require all the `compat.el` and` compat-xx.el `(xx means 24 to 28). 

I tried to require compat before building other packages, and it works well.  Let's take magit for example:
```
[submodule "magit"]
	no-byte-compile = lisp/magit-libgit.el
	build-step = (require 'compat)
	build-step = (require 'compat-24)
	build-step = (require 'compat-25)
	build-step = (require 'compat-26)
	build-step = (require 'compat-27)
	build-step = (require 'compat-28)
	build-step = borg-compile
        build-step = borg-update-autoloads
	path = lib/magit
	url = git@github.com:magit/magit.git
```

**Comment by Eason0210** [comment]
@tarsius  Updated : 
For `magit`, only need `(require 'compat-26)` and `(require 'compat-26)`
For `epkg`, only need `(require 'compat-26)` 
```
[submodule "epkg"]
	build-step = (require 'compat-26)
	build-step = borg-compile
        build-step = borg-update-autoloads
	path = lib/epkg
	url = git@github.com:emacscollective/epkg.git
```

**Comment by eklitzke** [comment]
FYI I also get this warning message in Spacemacs with Emacs head, and I'm not using borg.

**Comment by owinebar** [comment]
> FYI I also get this warning message in Spacemacs with Emacs head, and I'm not using borg.

I get it when "manually" batch compiling packaging.  Don't ask me why, but if you only native-compile the compat-26 and compat-27 packages you still get the error.  If you byte compile those files, there's no error, even if the ELN files are still in the load path for native compiled files.

**Comment by arichiardi** [comment]
FWIW compiling `compate` (with a custom `build-step`) before `magit` solved it for me ;)

in my `.gitmodules` I have got

```
[submodule "compat"]
	path = lib/compat
	url = git@github.com:emacsmirror/compat.git
	build-step = make
[submodule "magit"]
	path = lib/magit
	url = git@github.com:magit/magit.git
```

then I did

`make lib/compat lib/magit`

**Comment by tarsius** [comment]
You shouldn't have to take any extra precautions anymore.

I forgot to come back here and mention that this should be fixed now because of:

1. I added a workaround in https://github.com/emacscollective/borg/commit/a968c5e7eb771ca4d123b61f8058a1bf8e87bc11. (Well, `make clean lib/magit` would still fail, but `make clean build` wouldn't).
2. I also fixed it upstream in https://github.com/phikal/compat.el/commit/12774be87c7d52640be4c3d54e65236ba69b9a0e.
